llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -339,8 +339,8 @@ static Value *foldLogOpOfMaskedICmps_NotAllZeros_BMask_Mixed(
   //
   // We currently handle the case of B, C, D, E are constant.
   //
-  const APInt *BCst, *CCst, *DCst, *OrigECst;
-  if (!match(B, m_APInt(BCst)) || !match(C, m_APInt(CCst)) ||
+  const APInt *BCst, *DCst, *OrigECst;
+  if (!match(B, m_APInt(BCst)) || !match(C, m_Zero()) ||
```

Comment:
Was this just buggy before?

Reply:
It should be checked by the caller of `foldLogOpOfMaskedICmps_NotAllZeros_BMask_Mixed`. Can I use an assertion here instead?

Reply:
Not the easiest code to follow...  Doesn't seem obvious to me that `C` is enforced as 0 in the prelim code.

Reply:
See also https://github.com/llvm/llvm-project/blob/2a5f7e58d71c394b241fdd1d905041ad0537acab/llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp#L76-L114 and https://github.com/llvm/llvm-project/blob/2a5f7e58d71c394b241fdd1d905041ad0537acab/llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp#L446-L479

`Mask_NotAllZeros` stands for the pattern `icmp ne (A & B), 0`.

Reply:
The comments make sense but code like: https://github.com/llvm/llvm-project/blob/2a5f7e58d71c394b241fdd1d905041ad0537acab/llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp#L145 where we can seemingly end up w/ `Mask_NotAllZeros` with pow2 `C` are still unclear to me.

Reply:
`icmp eq (Pow2 & B), Pow2` can be canonicalized into `icmp ne (Pow2 & B), 0`. I think it is ok to remove the unused parameter `C`.

Reply:
Yeah, if we are potentially handling `icmp eq (Pow2 & B), Pow2` form of `ic.p ne (Pow2 & B), 0` (which agreed is fine) we shouldn't add an assert/

Reply:
Done.